File: course_4/module_2/examples/vertex_cover.go
package main

import (
	"fmt"
)

// Graph represents an undirected graph using an adjacency list.
type Graph map[string][]string

// VertexCover represents the vertex cover algorithm with memoization.
type VertexCover struct {
	memo map[string]bool // Memoization to store intermediate results.
}

// NewVertexCover initializes a VertexCover instance.
func NewVertexCover() *VertexCover {
	return &VertexCover{
		memo: make(map[string]bool),
	}
}

// VertexCoverSolver determines if there exists a vertex cover of size k or smaller.
func (vc *VertexCover) VertexCoverSolver(graph Graph, k int) bool {
	// Serialize the graph and k into a key for memoization.
	graphKey := vc.serializeGraph(graph) + fmt.Sprintf("_%d", k)

	// Check if the result is already memoized.
	if result, exists := vc.memo[graphKey]; exists {
		return result
	}

	// Base cases for recursion.
	if k < 0 {
		return false // Negative k is invalid.
	}
	if len(graph) == 0 {
		return true // Empty graph is trivially covered.
	}
	if k == 0 {
		return false // Non-empty graph cannot have a vertex cover of size 0.
	}

	// Pick a vertex with the highest degree (for better performance).
	var u string
	maxDegree := 0
	for vertex, neighbors := range graph {
		if len(neighbors) > maxDegree {
			u = vertex
			maxDegree = len(neighbors)
		}
	}

	// Select the first neighbor of u.
	v := graph[u][0]

	// Create subgraphs by removing u and v along with their incident edges.
	graphU := vc.createSubgraphExcludingVertices(graph, []string{u})
	graphV := vc.createSubgraphExcludingVertices(graph, []string{v})

	// Recursively check for vertex covers of size k-1.
	coverU := vc.VertexCoverSolver(graphU, k-1)
	coverV := vc.VertexCoverSolver(graphV, k-1)

	// Memoize and return the result.
	result := coverU || coverV
	vc.memo[graphKey] = result
	return result
}

// createSubgraphExcludingVertices creates a new subgraph by removing given vertices and their incident edges.
func (vc *VertexCover) createSubgraphExcludingVertices(graph Graph, vertices []string) Graph {
	// Create a set of vertices to remove for fast lookup.
	remove := make(map[string]bool)
	for _, v := range vertices {
		remove[v] = true
	}

	// Create a new graph excluding specified vertices and their edges.
	newGraph := make(Graph)
	for v, neighbors := range graph {
		if remove[v] {
			continue
		}
		filteredNeighbors := []string{}
		for _, neighbor := range neighbors {
			if !remove[neighbor] {
				filteredNeighbors = append(filteredNeighbors, neighbor)
			}
		}
		if len(filteredNeighbors) > 0 {
			newGraph[v] = filteredNeighbors
		}
	}
	return newGraph
}

// serializeGraph creates a string representation of the graph for memoization.
func (vc *VertexCover) serializeGraph(graph Graph) string {
	serialized := ""
	for v, neighbors := range graph {
		serialized += v + ":"
		for _, neighbor := range neighbors {
			serialized += neighbor + ","
		}
		serialized += ";"
	}
	return serialized
}

// Main function to test the implementation.
func main() {
	// Define a simple undirected graph as an adjacency list.
	graph := Graph{
		"A": {"B", "C"},
		"B": {"A", "D", "E"},
		"C": {"A", "F"},
		"D": {"B"},
		"E": {"B"},
		"F": {"C"},
	}

	// Create an instance of the VertexCover struct.
	vc := NewVertexCover()

	// Check if there is a vertex cover of size 2 or smaller.
	k := 2
	result := vc.VertexCoverSolver(graph, k)

	// Print the result.
	if result {
		fmt.Printf("Is there a vertex cover of size %d or smaller? Yes.\n", k)
	} else {
		fmt.Printf("Is there a vertex cover of size %d or smaller? No.\n", k)
	}
}

File: course_4/module_2/examples/vertex_cover.py
class VertexCover:
    """
    Class to represent the Vertex Cover problem and implement the recursive algorithm.
    """

    def __init__(self, graph):
        """
        Initialize the VertexCover class with an undirected graph.
        :param graph: Dictionary representing an undirected graph. Keys are vertices, and values are lists of adjacent vertices.
        """
        self.graph = graph
        self.memo = {}  # Memoization to avoid redundant computations.

    def vertex_cover(self, graph, k):
        """
        Determine if there exists a vertex cover of size k or smaller.
        :param graph: Dictionary representing the current graph.
        :param k: Integer, the size of the desired vertex cover.
        :return: True if a vertex cover of size k or smaller exists, otherwise False.
        """
        # Check memoized results.
        graph_key = tuple(sorted((v, tuple(sorted(neighbors))) for v, neighbors in graph.items()))
        if (graph_key, k) in self.memo:
            return self.memo[(graph_key, k)]

        # Base cases for recursion.
        if k < 0:
            return False
        if not graph:
            return True
        if k == 0:
            return False

        # Pick an arbitrary edge (u, v) from the graph.
        u = next(iter(graph))
        v = graph[u][0]

        # Create two subgraphs by removing u and v along with their incident edges.
        graph_u = self._create_subgraph_excluding_vertex(graph, u)
        graph_v = self._create_subgraph_excluding_vertex(graph, v)

        # Recursively check both subgraphs for vertex covers of size k-1.
        cover_u = self.vertex_cover(graph_u, k - 1)
        cover_v = self.vertex_cover(graph_v, k - 1)

        # Memoize and return the result.
        result = cover_u or cover_v
        self.memo[(graph_key, k)] = result
        return result

    @staticmethod
    def _create_subgraph_excluding_vertex(graph, vertex):
        """
        Create a new subgraph by removing a given vertex and its incident edges.
        :param graph: The current graph.
        :param vertex: The vertex to remove.
        :return: A new graph with the vertex and its edges removed.
        """
        new_graph = {v: [w for w in neighbors if w != vertex] for v, neighbors in graph.items() if v != vertex}
        return {v: neighbors for v, neighbors in new_graph.items() if neighbors}


# Example usage:
if __name__ == "__main__":
    # Define a simple undirected graph as a dictionary.
    graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D', 'E'],
        'C': ['A', 'F'],
        'D': ['B'],
        'E': ['B'],
        'F': ['C']
    }

    # Create an instance of the VertexCover class.
    vc = VertexCover(graph)

    # Check if there is a vertex cover of size 2 or smaller.
    k = 2
    result = vc.vertex_cover(graph, k)

    # Print the result.
    print(f"Is there a vertex cover of size {k} or smaller? {'Yes' if result else 'No'}.")
